File: services/common/company.py
"""Legal entity profile for customer-facing channels (voice, email, contracts)."""

from __future__ import annotations

import os
from dataclasses import asdict, dataclass
from typing import Any


@dataclass(frozen=True)
class CompanyProfile:
    legal_name: str
    short_name: str
    ogrn: str
    ogrn_date: str
    registration_date: str
    inn: str
    kpp: str
    legal_address: str
    director_title: str
    director_name: str
    director_since: str
    status: str
    status_assigned: str
    okved_code: str
    okved_name: str
    tax_authority: str
    tax_authority_since: str
    okpo: str
    okato: str
    oktmo: str
    okfs: str
    okfs_name: str
    okogu: str
    okogu_name: str
    okopf: str
    okopf_name: str
    phone: str = ""
    email: str = ""
    website: str = ""


DEFAULT_COMPANY = CompanyProfile(
    legal_name='ООО «ЖД Трансинвест»',
    short_name="ЖД Трансинвест",
    ogrn="1205400060260",
    ogrn_date="16.12.2020",
    registration_date="16.12.2020",
    inn="5401401513",
    kpp="540601001",
    legal_address=(
        "630007, Новосибирская область, г Новосибирск, "
        "Октябрьская ул, зд. 42, офис 308"
    ),
    director_title="Директор",
    director_name="Тарасова Александра Викторовна",
    director_since="16.12.2020",
    status="микропредприятие",
    status_assigned="10.01.2021",
    okved_code="52.29",
    okved_name="Деятельность вспомогательная прочая, связанная с перевозками",
    tax_authority=(
        "Межрайонная инспекция ФНС России № 22 по Новосибирской области"
    ),
    tax_authority_since="23.08.2022",
    okpo="46701699",
    okato="50401386000",
    oktmo="50701000001",
    okfs="16",
    okfs_name="Частная собственность",
    okogu="4210014",
    okogu_name=(
        "Организации, учрежденные юридическими лицами или гражданами, "
        "или юридическими лицами и гражданами совместно"
    ),
    okopf="12300",
    okopf_name="Общества с ограниченной ответственностью",
)


def _env(key: str, fallback: str) -> str:
    v = os.getenv(key)
    return v.strip() if v and v.strip() else fallback


def load_company() -> CompanyProfile:
    d = DEFAULT_COMPANY
    return CompanyProfile(
        legal_name=_env("COMPANY_LEGAL_NAME", d.legal_name),
        short_name=_env("COMPANY_SHORT_NAME", d.short_name),
        ogrn=_env("COMPANY_OGRN", d.ogrn),
        ogrn_date=_env("COMPANY_OGRN_DATE", d.ogrn_date),
        registration_date=_env("COMPANY_REGISTRATION_DATE", d.registration_date),
        inn=_env("COMPANY_INN", d.inn),
        kpp=_env("COMPANY_KPP", d.kpp),
        legal_address=_env("COMPANY_LEGAL_ADDRESS", d.legal_address),
        director_title=_env("COMPANY_DIRECTOR_TITLE", d.director_title),
        director_name=_env("COMPANY_DIRECTOR_NAME", d.director_name),
        director_since=_env("COMPANY_DIRECTOR_SINCE", d.director_since),
        status=_env("COMPANY_STATUS", d.status),
        status_assigned=_env("COMPANY_STATUS_ASSIGNED", d.status_assigned),
        okved_code=_env("COMPANY_OKVED_CODE", d.okved_code),
        okved_name=_env("COMPANY_OKVED_NAME", d.okved_name),
        tax_authority=_env("COMPANY_TAX_AUTHORITY", d.tax_authority),
        tax_authority_since=_env(
            "COMPANY_TAX_AUTHORITY_SINCE", d.tax_authority_since
        ),
        okpo=_env("COMPANY_OKPO", d.okpo),
        okato=_env("COMPANY_OKATO", d.okato),
        oktmo=_env("COMPANY_OKTMO", d.oktmo),
        okfs=_env("COMPANY_OKFS", d.okfs),
        okfs_name=_env("COMPANY_OKFS_NAME", d.okfs_name),
        okogu=_env("COMPANY_OKOGU", d.okogu),
        okogu_name=_env("COMPANY_OKOGU_NAME", d.okogu_name),
        okopf=_env("COMPANY_OKOPF", d.okopf),
        okopf_name=_env("COMPANY_OKOPF_NAME", d.okopf_name),
        phone=_env("COMPANY_PHONE", d.phone),
        email=_env("COMPANY_EMAIL", d.email),
        website=_env("COMPANY_WEBSITE", d.website),
    )


def company_display_name(company: CompanyProfile | None = None) -> str:
    c = company or load_company()
    voice = os.getenv("VOICE_COMPANY_NAME", "").strip()
    return voice or c.short_name or c.legal_name


def company_as_dict(company: CompanyProfile | None = None) -> dict[str, Any]:
    return asdict(company or load_company())


def company_requisites_md(company: CompanyProfile | None = None) -> str:
    c = company or load_company()
    lines = [
        f"**{c.legal_name}**",
        f"ОГРН {c.ogrn} от {c.ogrn_date}",
        f"Дата регистрации: {c.registration_date}",
        f"ИНН/КПП: {c.inn} / {c.kpp}",
        f"Юридический адрес: {c.legal_address}",
        (
            f"Руководитель: {c.director_title} {c.director_name} "
            f"(с {c.director_since})"
        ),
        f"Статус: {c.status} (присвоен {c.status_assigned})",
        f"ОКВЭД: {c.okved_code} — {c.okved_name}",
        f"Налоговый орган: {c.tax_authority} (с {c.tax_authority_since})",
        f"ОКПО {c.okpo}; ОКАТО {c.okato}; ОКТМО {c.oktmo}",
        (
            f"ОКФС {c.okfs} ({c.okfs_name}); ОКОГУ {c.okogu}; "
            f"ОКОПФ {c.okopf} ({c.okopf_name})"
        ),
    ]
    if c.phone:
        lines.append(f"Тел.: {c.phone}")
    if c.email:
        lines.append(f"Email: {c.email}")
    if c.website:
        lines.append(f"Сайт: {c.website}")
    return "\n".join(lines)

File: packages/shared/src/company.ts
/**
 * Legal entity profile for customer-facing channels (voice, email, contracts).
 * Override via COMPANY_* env vars. Product name remains AutoLogistics OS.
 */
export interface CompanyProfile {
  legal_name: string;
  short_name: string;
  ogrn: string;
  ogrn_date: string;
  registration_date: string;
  inn: string;
  kpp: string;
  legal_address: string;
  director_title: string;
  director_name: string;
  director_since: string;
  status: string;
  status_assigned: string;
  okved_code: string;
  okved_name: string;
  tax_authority: string;
  tax_authority_since: string;
  okpo: string;
  okato: string;
  oktmo: string;
  okfs: string;
  okfs_name: string;
  okogu: string;
  okogu_name: string;
  okopf: string;
  okopf_name: string;
  phone: string;
  email: string;
  website: string;
}

export const DEFAULT_COMPANY: CompanyProfile = {
  legal_name: 'ООО «ЖД Трансинвест»',
  short_name: "ЖД Трансинвест",
  ogrn: "1205400060260",
  ogrn_date: "16.12.2020",
  registration_date: "16.12.2020",
  inn: "5401401513",
  kpp: "540601001",
  legal_address:
    "630007, Новосибирская область, г Новосибирск, Октябрьская ул, зд. 42, офис 308",
  director_title: "Директор",
  director_name: "Тарасова Александра Викторовна",
  director_since: "16.12.2020",
  status: "микропредприятие",
  status_assigned: "10.01.2021",
  okved_code: "52.29",
  okved_name: "Деятельность вспомогательная прочая, связанная с перевозками",
  tax_authority:
    "Межрайонная инспекция ФНС России № 22 по Новосибирской области",
  tax_authority_since: "23.08.2022",
  okpo: "46701699",
  okato: "50401386000",
  oktmo: "50701000001",
  okfs: "16",
  okfs_name: "Частная собственность",
  okogu: "4210014",
  okogu_name:
    "Организации, учрежденные юридическими лицами или гражданами, или юридическими лицами и гражданами совместно",
  okopf: "12300",
  okopf_name: "Общества с ограниченной ответственностью",
  phone: "",
  email: "",
  website: "",
};

function env(key: string, fallback: string): string {
  const v = process.env[key];
  return v != null && v.trim() !== "" ? v.trim() : fallback;
}

/** Load company profile from COMPANY_* env (falls back to DEFAULT_COMPANY). */
export function loadCompanyFromEnv(
  envMap: NodeJS.ProcessEnv = process.env
): CompanyProfile {
  const get = (key: string, fallback: string) => {
    const v = envMap[key];
    return v != null && v.trim() !== "" ? v.trim() : fallback;
  };
  return {
    legal_name: get("COMPANY_LEGAL_NAME", DEFAULT_COMPANY.legal_name),
    short_name: get("COMPANY_SHORT_NAME", DEFAULT_COMPANY.short_name),
    ogrn: get("COMPANY_OGRN", DEFAULT_COMPANY.ogrn),
    ogrn_date: get("COMPANY_OGRN_DATE", DEFAULT_COMPANY.ogrn_date),
    registration_date: get(
      "COMPANY_REGISTRATION_DATE",
      DEFAULT_COMPANY.registration_date
    ),
    inn: get("COMPANY_INN", DEFAULT_COMPANY.inn),
    kpp: get("COMPANY_KPP", DEFAULT_COMPANY.kpp),
    legal_address: get("COMPANY_LEGAL_ADDRESS", DEFAULT_COMPANY.legal_address),
    director_title: get("COMPANY_DIRECTOR_TITLE", DEFAULT_COMPANY.director_title),
    director_name: get("COMPANY_DIRECTOR_NAME", DEFAULT_COMPANY.director_name),
    director_since: get("COMPANY_DIRECTOR_SINCE", DEFAULT_COMPANY.director_since),
    status: get("COMPANY_STATUS", DEFAULT_COMPANY.status),
    status_assigned: get(
      "COMPANY_STATUS_ASSIGNED",
      DEFAULT_COMPANY.status_assigned
    ),
    okved_code: get("COMPANY_OKVED_CODE", DEFAULT_COMPANY.okved_code),
    okved_name: get("COMPANY_OKVED_NAME", DEFAULT_COMPANY.okved_name),
    tax_authority: get("COMPANY_TAX_AUTHORITY", DEFAULT_COMPANY.tax_authority),
    tax_authority_since: get(
      "COMPANY_TAX_AUTHORITY_SINCE",
      DEFAULT_COMPANY.tax_authority_since
    ),
    okpo: get("COMPANY_OKPO", DEFAULT_COMPANY.okpo),
    okato: get("COMPANY_OKATO", DEFAULT_COMPANY.okato),
    oktmo: get("COMPANY_OKTMO", DEFAULT_COMPANY.oktmo),
    okfs: get("COMPANY_OKFS", DEFAULT_COMPANY.okfs),
    okfs_name: get("COMPANY_OKFS_NAME", DEFAULT_COMPANY.okfs_name),
    okogu: get("COMPANY_OKOGU", DEFAULT_COMPANY.okogu),
    okogu_name: get("COMPANY_OKOGU_NAME", DEFAULT_COMPANY.okogu_name),
    okopf: get("COMPANY_OKOPF", DEFAULT_COMPANY.okopf),
    okopf_name: get("COMPANY_OKOPF_NAME", DEFAULT_COMPANY.okopf_name),
    phone: get("COMPANY_PHONE", DEFAULT_COMPANY.phone),
    email: get("COMPANY_EMAIL", DEFAULT_COMPANY.email),
    website: get("COMPANY_WEBSITE", DEFAULT_COMPANY.website),
  };
}

/** Display name for voice / email From / signatures. */
export function companyDisplayName(
  company: CompanyProfile = loadCompanyFromEnv()
): string {
  return (
    env("VOICE_COMPANY_NAME", "") ||
    company.short_name ||
    company.legal_name
  );
}

/** Markdown block for contract drafts / legal context. */
export function companyRequisitesMd(
  company: CompanyProfile = loadCompanyFromEnv()
): string {
  const lines = [
    `**${company.legal_name}**`,
    `ОГРН ${company.ogrn} от ${company.ogrn_date}`,
    `Дата регистрации: ${company.registration_date}`,
    `ИНН/КПП: ${company.inn} / ${company.kpp}`,
    `Юридический адрес: ${company.legal_address}`,
    `Руководитель: ${company.director_title} ${company.director_name} (с ${company.director_since})`,
    `Статус: ${company.status} (присвоен ${company.status_assigned})`,
    `ОКВЭД: ${company.okved_code} — ${company.okved_name}`,
    `Налоговый орган: ${company.tax_authority} (с ${company.tax_authority_since})`,
    `ОКПО ${company.okpo}; ОКАТО ${company.okato}; ОКТМО ${company.oktmo}`,
    `ОКФС ${company.okfs} (${company.okfs_name}); ОКОГУ ${company.okogu}; ОКОПФ ${company.okopf} (${company.okopf_name})`,
  ];
  if (company.phone) lines.push(`Тел.: ${company.phone}`);
  if (company.email) lines.push(`Email: ${company.email}`);
  if (company.website) lines.push(`Сайт: ${company.website}`);
  return lines.join("\n");
}
